<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Description &mdash; franklin v0.1 documentation</title>
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '#',
        VERSION:     '0.1',
        COLLAPSE_MODINDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="franklin v0.1 documentation" href="index.html" />
    <link rel="prev" title="Welcome to franklin’s documentation!" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="index.html" title="Welcome to franklin’s documentation!"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">franklin v0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="description">
<h1>Description<a class="headerlink" href="#description" title="Permalink to this headline">¶</a></h1>
<p>franklin is a bioinformatic application created to work on sequence analysis by using NGS (Next Generation Sequencing)  and sanger sequences. It is capable of cleaning reads, do de novo assembly or mapping against a reference and annotate SNPs, SSRs, ORFs, GO terms and sequence descriptions.
Our laboratory is focused on transcriptomic analysis, so the tool has been used and tested on transcriptomes. Some analyses will be useful for genome analysis, but other might not be appropriate.</p>
<p>For the analyses, in most cases, franklin uses external software like: <a class="reference external" href="http://sourceforge.net/apps/mediawiki/mira-assembler">mira</a>, <a class="reference external" href="http://bio-bwa.sourceforge.net/">bwa</a>, <a class="reference external" href="http://samtools.sourceforge.net/">samtools</a>, <a class="reference external" href="http://picard.sourceforge.net/index.shtml">picard</a>, etc.</p>
<p>The application works on Linux so experience with that operating system is advisable when using it.</p>
</div>
<div class="section" id="philosophy">
<h1>Philosophy<a class="headerlink" href="#philosophy" title="Permalink to this headline">¶</a></h1>
<p>franklin is an easy to use application capable of doing in a reliable way some NGS analyses. The main design directives have been:</p>
<blockquote>
<ul class="simple">
<li>analyses reproducibility</li>
<li>ease of use</li>
<li>modularity</li>
<li>standard format output</li>
</ul>
</blockquote>
<p>By using franklin we can run a complete analysis in a reproducible way. Every analysis parameter is configured in a text file and a log file is generated as the analysis progresses. Also the analysis is easy to do, instead of running multiple scripts and programs to do an analysis only one command is required.</p>
<p>An application like that in a fast moving field, as the NGS is, has the risk of stagnate rapidly. To avoid this pitfall special care has been taken in the design of the architecture of franklin. Everything within the application is modular and several independent layers have been differentiated within the code to facilitate the maintenance of the code. In fact before its public release franklin has gone through refactorings of several modules that have not affected the overall structure of the application.</p>
</div>
<div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p>To install franklin python 2.6 and <a class="reference external" href="http://biopython.org/wiki/Main_Page">Biopython</a> are required. python is installed by default in the usual Linux distributions but you should check the version. Installing <a class="reference external" href="http://biopython.org/wiki/Main_Page">Biopython</a> is as easy as installing franklin. Once the tarball is downloaded you have to run the following command:</p>
<div class="highlight-python"><pre>$ python setup.py install</pre>
</div>
<p>franklin requires several external tools to run the analyses. So before running the analyses please install them.</p>
<table border="1" class="docutils">
<colgroup>
<col width="26%" />
<col width="74%" />
</colgroup>
<thead valign="bottom">
<tr><th class="head">analysis</th>
<th class="head">external tools required</th>
</tr>
</thead>
<tbody valign="top">
<tr><td>read cleaning</td>
<td><a class="reference external" href="http://lucy.sourceforge.net/">lucy</a>, <a class="reference external" href="http://www.ebi.ac.uk/~guy/exonerate/">exonerate</a>, <a class="reference external" href="http://web.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">blast</a>, <a href="#id1"><span class="problematic" id="id2">Univec_</span></a> (database), <a class="reference external" href="http://compbio.dfci.harvard.edu/tgi/software/">mdust</a>, <a class="reference external" href="http://compbio.dfci.harvard.edu/tgi/software/">trimpoly</a></td>
</tr>
<tr><td>assembly</td>
<td><a class="reference external" href="http://sourceforge.net/apps/mediawiki/mira-assembler">mira</a></td>
</tr>
<tr><td>mapping</td>
<td><a class="reference external" href="http://bio-bwa.sourceforge.net/">bwa</a>, <a class="reference external" href="http://samtools.sourceforge.net/">samtools</a>, <a class="reference external" href="http://picard.sourceforge.net/index.shtml">picard</a></td>
</tr>
<tr><td>bam realignment</td>
<td><a class="reference external" href="http://www.broadinstitute.org/gsa/wiki/index.php/The_Genome_Analysis_Toolkit">GATK</a></td>
</tr>
<tr><td>SNP calling</td>
<td><a class="reference external" href="http://code.google.com/p/pysam/">pysam</a></td>
</tr>
<tr><td>ORF annotation</td>
<td><a class="reference external" href="http://estscan.sourceforge.net/">ESTScan</a></td>
</tr>
<tr><td>ortholog annotation</td>
<td><a class="reference external" href="http://web.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">blast</a></td>
</tr>
<tr><td>description annotation</td>
<td><a class="reference external" href="http://web.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">blast</a></td>
</tr>
<tr><td>SSR annotation</td>
<td>sputnik</td>
</tr>
<tr><td>cdna intron annotation</td>
<td><a class="reference external" href="http://web.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">blast</a>, <a class="reference external" href="http://emboss.sourceforge.net/">emboss</a></td>
</tr>
<tr><td>GO annotation</td>
<td><a class="reference external" href="http://web.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">blast</a>, <a class="reference external" href="http://www.blast2go.org/">blast2go</a></td>
</tr>
</tbody>
</table>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <h3><a href="index.html">Table Of Contents</a></h3>
            <ul>
<li><a class="reference external" href="#">Description</a></li>
<li><a class="reference external" href="#philosophy">Philosophy</a></li>
<li><a class="reference external" href="#installation">Installation</a></li>
</ul>

            <h4>Previous topic</h4>
            <p class="topless"><a href="index.html"
                                  title="previous chapter">Welcome to franklin&#8217;s documentation!</a></p>
            <h3>This Page</h3>
            <ul class="this-page-menu">
              <li><a href="_sources/introduction.txt"
                     rel="nofollow">Show Source</a></li>
            </ul>
          <div id="searchbox" style="display: none">
            <h3>Quick search</h3>
              <form class="search" action="search.html" method="get">
                <input type="text" name="q" size="18" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
              </form>
              <p class="searchtip" style="font-size: 90%">
              Enter search terms or a module, class or function name.
              </p>
          </div>
          <script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="index.html" title="Welcome to franklin’s documentation!"
             >previous</a> |</li>
        <li><a href="index.html">franklin v0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
      &copy; Copyright 2010, Jose Blanca.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 0.6.5.
    </div>
  </body>
</html>