

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Cleaning sequence reads &mdash; ngs_backbone v0.1 documentation</title>
    <link rel="stylesheet" href="_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="ngs_backbone v0.1 documentation" href="index.html" />
    <link rel="next" title="Mira assembly" href="assembly.html" />
    <link rel="prev" title="Installation" href="install.html" /> 
  </head>
  <body>
    <div class="header-wrapper">
      <div class="header">
        <h1><a href="index.html">ngs_backbone v0.1 documentation</a></h1>
        <div class="rel">
          <a href="install.html" title="Installation"
             accesskey="P">previous</a> |
          <a href="assembly.html" title="Mira assembly"
             accesskey="N">next</a> |
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="cleaning-sequence-reads">
<span id="clean-reads"></span><h1>Cleaning sequence reads<a class="headerlink" href="#cleaning-sequence-reads" title="Permalink to this headline">¶</a></h1>
<p>ngs_backbone can clean sanger, 454, illumina and solid sequences.
This process usually involves vector and adaptor removal, bad quality regions trimming and short sequence filtering.
There are four cleaning pipelines defined in ngs_backbone that are used depending on the platform and on the quality availability:</p>
<dl class="docutils">
<dt>long reads with quality</dt>
<dd>for sanger and 454 sequences with quality information</dd>
<dt>long reads without quality</dt>
<dd>for sanger reads without quality information</dd>
<dt>solexa</dt>
<dd>for short illumina reads</dd>
<dt>solid</dt>
<dd>for solid reads</dd>
</dl>
<p>A collection of cleaning steps are available that compose each one of these pipelines.
These steps are:</p>
<dl class="docutils">
<dt>adaptor removal</dt>
<dd>Each sequence is align against the adaptors found in a fasta file. The external tool used to do the matching is blast with its short sequence algorithm. If a match is found this section of the read is removed.</dd>
<dt>short adaptor removal</dt>
<dd>ngs_backbone will look for adaptors shorter than 15 bp with exact matches.</dd>
<dt>precise vector removal</dt>
<dd>If the vector and cloning site is known lucy can be used to remove the vector in a precise way.</dd>
<dt>vector removal</dt>
<dd>If only the vector is known we can use blast to remove it.</dd>
<dt>general vector removal</dt>
<dd>The reads are compared against the Univec database using blast to look for remaining vectors.</dd>
<dt>bad quality trimming</dt>
<dd>There are two algorithms used to remove the bad quality sequence extremes. If the sequence is long lucy (454 and sanger) is used for this task otherwise ngs_backbone does the job (illumina and solid).</dd>
<dt>bad quality trimming by Ns</dt>
<dd>When there are no qualities available we can infer the quality from the percent of Ns found in the sequence. This step removes regions with a high number of Ns.</dd>
<dt>low complexity masking</dt>
<dd>The regions with a low complexity can be masked by using mdust</dd>
<dt>edge removal</dt>
<dd>After all the other modules are run we can delete a fixed amount of bases from the sequence extremes</dd>
<dt>short sequence filtering</dt>
<dd>When the process for one sequence is completed a minimum length criteria is applied.</dd>
</dl>
<p>The pipelines are:</p>
<dl class="docutils">
<dt>long reads with quality</dt>
<dd>adaptor removal, precise vector removal, bad quality trimming, general vector removal, low complexity masking, word removal, edge removal, and short sequence filtering</dd>
<dt>long reads without quality</dt>
<dd>general vector removal, bad quality trimming by Ns, low complexity masking, word removal, edge removal, and short sequence filtering</dd>
<dt>solexa</dt>
<dd>adaptor removal, bad quality trimming,  and short sequence filtering</dd>
<dt>solid</dt>
<dd>bad quality trimming and short sequence filtering</dd>
</dl>
<div class="section" id="input-and-output-files">
<h2>Input and output files<a class="headerlink" href="#input-and-output-files" title="Permalink to this headline">¶</a></h2>
<p>The reads to be cleaned should be in the project directory under /reads/raw/.
The <a class="reference internal" href="introduction.html"><em>naming conventions</em></a> should be followed by these files, especially the bit regarding to the extension and the platform.
The output files will have the same names, but they will be located at /reads/cleaned/.
The analysis will proceed for all sequence files found in /reads/raw, if a matching file is not found in /reads/cleaned/ a new cleaned file will be generated.
If a matching file is found in /reads/cleaned/ these file will not be overwritten, so the analysis for this file will not be repeated until the file from /reads/cleaned is removed.</p>
</div>
<div class="section" id="configuration-parameters">
<span id="clean-config"></span><h2>Configuration parameters<a class="headerlink" href="#configuration-parameters" title="Permalink to this headline">¶</a></h2>
<p>The configuration for the cleaning analysis is found in the Cleaning section on the ngs_backbone.conf file. The parameters are:</p>
<dl class="docutils">
<dt>vector_database</dt>
<dd>The blast database that will be used to look for clonning vectors.</dd>
<dt>adaptors_file_454</dt>
<dd>A path to a fasta file containing the adaptors used to build the 454 library. They will be removed from the cleanend reads.</dd>
<dt>adaptors_file_sanger</dt>
<dd>Idem for the sanger sequences</dd>
<dt>adaptors_file_illumina</dt>
<dd>Idem for the illumina sequences</dd>
<dt>short_adaptors_454</dt>
<dd>A list of words to be removed. They can be regular expressions.</dd>
<dt>short_adaptors_sanger</dt>
<dd>Idem for the sanger sequences</dd>
<dt>short_adaptors_illumina</dt>
<dd>Idem for the illumina sequences</dd>
<dt>edge_removal -&gt; 454_left</dt>
<dd>A fixed number of bases to be removed from the left edge of the 454 reads.</dd>
<dt>edge_removal -&gt; 454_right</dt>
<dd>Idem for the right edge of the 454 reads</dd>
<dt>edge_removal -&gt; sanger_left</dt>
<dd>Idem for the left edge of the sanger reads</dd>
<dt>edge_removal -&gt; sanger_right</dt>
<dd>Idem for the right edge of the sanger reads</dd>
<dt>edge_removal -&gt; illumina_left</dt>
<dd>Idem for the left edge of the illumina reads</dd>
<dt>edge_removal -&gt; illumina_right</dt>
<dd>Idem for the right edge of the illumina reads</dd>
<dt>strip_n_percent</dt>
<dd>Threshold used for the trimming of regions with a lot of Ns in sequences with no quality information available. Lower values (e.g. 1.5) are more stringent.</dd>
<dt>min_seq_length</dt>
<dd>The minimum sequence length allowable after the cleaning is done. All sequences shorter than these values will be discarded. This is a subsection with one value for each platform 454, sanger and illumina.</dd>
<dt>lucy -&gt; vector_settings</dt>
<dd>A path to a lucy settings file with the splice and vectors files to be used by lucy</dd>
<dt>lucy -&gt;bracket</dt>
<dd>Look at lucy man page before changing defaults.</dd>
<dt>lucy -&gt; window</dt>
<dd>Look at lucy man page before changing defaults.</dd>
<dt>lucy -&gt; error</dt>
<dd>Look at lucy man page before changing defaults.</dd>
</dl>
</div>
<div class="section" id="lucy-settings">
<h2>lucy settings<a class="headerlink" href="#lucy-settings" title="Permalink to this headline">¶</a></h2>
<p>The lucy settings file should have the following format:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="p">{</span><span class="s">&#39;library1&#39;</span><span class="p">:{</span><span class="s">&#39;vector_file&#39;</span><span class="p">:</span><span class="s">&#39;lib1_vector.fasta&#39;</span><span class="p">,</span> <span class="s">&#39;splice_file&#39;</span><span class="p">:</span><span class="s">&#39;lib1_splice.fasta&#39;</span><span class="p">},</span>
 <span class="s">&#39;library2&#39;</span><span class="p">:{</span><span class="s">&#39;vector_file&#39;</span><span class="p">:</span><span class="s">&#39;lib2_vector.fasta&#39;</span><span class="p">,</span> <span class="s">&#39;splice_file&#39;</span><span class="p">:</span><span class="s">&#39;lib2_splice.fasta&#39;</span><span class="p">},}</span>
</pre></div>
</div>
<p>In this file the paths to the vector and splice files for lucy should be stated for every library to be cleaned by lucy.
The library name will be scraped from the read sequence file (that should follow the <a class="reference internal" href="introduction.html"><em>naming conventions</em></a>.
The vector file is just a fasta file, the information to be set in the splice file should is explained in the lucy man page.</p>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          <h3>Table Of Contents</h3>
          <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#third-party-software">Third party software</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#usage">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#naming-conventions">Naming conventions</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#available-analyses">Available analyses</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#parallel-operation">Parallel operation</a></li>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Cleaning sequence reads</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#input-and-output-files">Input and output files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#configuration-parameters">Configuration parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#lucy-settings">lucy settings</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="assembly.html">Mira assembly</a></li>
<li class="toctree-l1"><a class="reference internal" href="mapping.html">Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="mapping.html#bam-realignment">Bam realignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="annotation.html">Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="snv_filters.html">Snv filters</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="seqio.html">seq_io</a></li>
<li class="toctree-l1"><a class="reference internal" href="seqio.html#seqio-supported-formats">Seqio supported formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="architecture.html">Architecture</a></li>
<li class="toctree-l1"><a class="reference internal" href="download.html">Download</a></li>
</ul>

          <h3 style="margin-top: 1.5em;">Search</h3>
          <form class="search" action="search.html" method="get">
            <input type="text" name="q" size="18" />
            <input type="submit" value="Go" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
          </form>
          <p class="searchtip" style="font-size: 90%">
            Enter search terms or a module, class or function name.
          </p>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <a href="install.html" title="Installation"
             >previous</a> |
          <a href="assembly.html" title="Mira assembly"
             >next</a> |
          <a href="genindex.html" title="General Index"
             >index</a>
            <br/>
            <a href="_sources/cleaning.txt"
               rel="nofollow">Show Source</a>
        </div>

        <div class="right">
          
    <div class="footer">
        &copy; Copyright 2010, Jose Blanca.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.7.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>