

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Installation &mdash; ngs_backbone v0.1 documentation</title>
    <link rel="stylesheet" href="_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="ngs_backbone v0.1 documentation" href="index.html" />
    <link rel="next" title="Cleaning sequence reads" href="cleaning.html" />
    <link rel="prev" title="Introduction" href="introduction.html" /> 
  </head>
  <body>
    <div class="header-wrapper">
      <div class="header">
        <h1><a href="index.html">ngs_backbone v0.1 documentation</a></h1>
        <div class="rel">
          <a href="introduction.html" title="Introduction"
             accesskey="P">previous</a> |
          <a href="cleaning.html" title="Cleaning sequence reads"
             accesskey="N">next</a> |
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p>This is a summary on how to install ngs_backbone, below you have a <a class="reference internal" href="#step-by-step"><em>detailed explanation</em></a>.</p>
<p>If you just want to check out how ngs_backbone works you could also download a complete <a class="reference external" href="http://www.virtualbox.org/">VirtualBox</a> <a class="reference external" href="http://bioinf.comav.upv.es/_downloads/ngs_machine_v2.tar.gz">virtual machine</a> already pre-installed with the software and with the files required to do our <a class="reference external" href="http://bioinf.comav.upv.es/courses/ngs_workshop/">NGS workshop</a>.
Although this machine will not be powerful enough to run most real sized analyses it can be an easy way to get a felling of the analysis process without spending any time at all setting up the environment.</p>
<p>To install ngs_backbond python 2.6 or better is required.
Also you need the python libraries <a class="reference external" href="http://biopython.org/wiki/Main_Page">Biopython</a>, and <a class="reference external" href="http://pypi.python.org/pypi/configobj/">ConfigObj</a>.
An easy way to install these libraries would be to use the packaging system of your Linux distribution or alternatively just run the following commands:</p>
<div class="highlight-python"><pre>$ easy_instal biopython
$ easy_install configobj</pre>
</div>
<p>If you don&#8217;t have easy_install installed in your computer.
You can grab it from here: <a class="reference external" href="http://pypi.python.org/pypi/setuptools">setuptools</a>.
Other optional, but highly recommended, dependencies are: <a class="reference external" href="http://bioinf.comav.upv.es/psubprocess/">psubprocess</a>, <a class="reference external" href="http://code.google.com/p/pysam/">pysam</a> and <a class="reference external" href="http://matplotlib.sourceforge.net/">matplotlib</a>.
All of these libraries are installed in the same way:</p>
<div class="highlight-python"><pre>$ tar -zxvf package.tar.gz
$ cd package
$ python setup.py install</pre>
</div>
<p>Most of the third party tools required by ngs_backbone are included in it, but if you&#8217;re planing to annotate GO terms you will have to install b2g4pipe(<a class="reference external" href="http://www.blast2go.org/">blast2go</a>).</p>
<p>For the vector cleaning the <a class="reference external" href="http://www.ncbi.nlm.nih.gov/VecScreen/UniVec.html">Univec</a> database can be used.
If you configure the cleaning process to use it you also have to install it.</p>
<div class="section" id="step-by-step-installation-instructions">
<span id="step-by-step"></span><h2>Step by step installation instructions<a class="headerlink" href="#step-by-step-installation-instructions" title="Permalink to this headline">¶</a></h2>
<div class="section" id="python-tools">
<h3>python tools<a class="headerlink" href="#python-tools" title="Permalink to this headline">¶</a></h3>
<p>ngs_backbone requires python2.6.
Chance are that it is already included in your distribution, but if it is not your distribution you can <a class="reference external" href="http://www.python.org/download/releases/">download</a> it and install it.
The other requirements are Python libraries.
Biopython and ConfigObj are required and pysam, psubprocess and matplotlib are optional.</p>
<p>It is quite possible that your distribution already includes Biopython and ConfigObj so you can use your package manager to install them, but we are going to explain here the manual process.
The install process is simple, you just have to download a bunch of python tools, unpack them and run &#8220;python2.6 install&#8221; on them.</p>
<p>To install Biopython you need <a class="reference external" href="http://numpy.scipy.org/">Numpy</a>. Download <a class="reference external" href="http://biopython.org/wiki/Main_Page">Biopython</a> and install it.</p>
<dl class="docutils">
<dt>::</dt>
<dd><p class="first">$ tar -xvzf numpy-1.4.1.tar.gz
$ cd numpy-1.4.1</p>
<p class="last">$ tar -xvzf biopython-1.54.tar.gz
$ cd biopython-1.54
$ python2.6 setup.py install</p>
</dd>
</dl>
<p>Download <a class="reference external" href="http://pypi.python.org/pypi/configobj/">ConfigObj</a> and install it.</p>
<div class="highlight-python"><pre>$ unzip configobj-4.7.1.zip
$ cd configobj-4.7.1
$ python2.6 setup.py install</pre>
</div>
<p>Download and install <a class="reference external" href="http://bioinf.comav.upv.es/psubprocess/">psubprocess</a> (If you don&#8217;t install it you won&#8217;t be able to run ngs_backbone in parallel).</p>
<div class="highlight-python"><pre>$ tar -xvzf psubprocess.0.1.1.tar.gz
$ cd psubprocess.0.1.1
$ python2.6 setup.py install</pre>
</div>
<p>To call the SNP you will need the library <a class="reference external" href="http://code.google.com/p/pysam/">pysam</a>.</p>
<div class="highlight-python"><pre>$ tar -xvzf pysam-0.4.1.tar.gz
$ cd pysam-0.4.1
$ python2.6 setup.py install</pre>
</div>
<p>To create the charts for the statistics you will need <a class="reference external" href="http://matplotlib.sourceforge.net/">matplotlib</a>.</p>
<div class="highlight-python"><pre>$ tar -xvzf matplotlib-1.0.1.tar.gz
$ cd matplotlib-1.0.1
$ python2.6 setup.py install</pre>
</div>
<p>Once we have it all we can install ngs_backbone.</p>
<div class="highlight-python"><pre>$ tar -xvzf ngs_backbone.0.2.0.tar.gz
$ cd ngs_backbone.0.2.0
$ python2.6 setup.py install</pre>
</div>
</div>
<div class="section" id="java-tools">
<h3>Java tools<a class="headerlink" href="#java-tools" title="Permalink to this headline">¶</a></h3>
<p>Three java tools are used: <a class="reference external" href="http://picard.sourceforge.net/index.shtml">picard</a> ,GATK_ and <a class="reference external" href="http://www.blast2go.org/">blast2go</a>.
The first two are provided with ngs_backbone so you don&#8217;t need to worry about them.</p>
<p>To install <a class="reference external" href="http://www.blast2go.org/">blast2go</a>, just download it and unpack it.</p>
<div class="highlight-python"><pre>$ tar zxvf b2g4pipe235.tar.gz
$ mv b2g4pipe235 /usr/local
$ updatedb</pre>
</div>
<p>It is advisable to run updatedb after setting everything to ease the ngs_backbone configuration.</p>
<p>Finally to run the java tools you need java.
If you don&#8217;t have it already installed just do.</p>
<div class="highlight-python"><pre>$ apt-get install sun-java6-jre</pre>
</div>
<p>After installing the whole pipeline you can run the <a class="reference external" href="http://bioinf.comav.upv.es/courses/ngs_workshop/">NGS workshop tutorial</a> to test the whole system.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          <h3>Table Of Contents</h3>
          <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#third-party-software">Third party software</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#usage">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#naming-conventions">Naming conventions</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#available-analyses">Available analyses</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#parallel-operation">Parallel operation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#step-by-step-installation-instructions">Step by step installation instructions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="cleaning.html">Cleaning sequence reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="assembly.html">Mira assembly</a></li>
<li class="toctree-l1"><a class="reference internal" href="mapping.html">Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="mapping.html#bam-realignment">Bam realignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="annotation.html">Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="snv_filters.html">Snv filters</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="seqio.html">seq_io</a></li>
<li class="toctree-l1"><a class="reference internal" href="seqio.html#seqio-supported-formats">Seqio supported formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="architecture.html">Architecture</a></li>
<li class="toctree-l1"><a class="reference internal" href="download.html">Download</a></li>
</ul>

          <h3 style="margin-top: 1.5em;">Search</h3>
          <form class="search" action="search.html" method="get">
            <input type="text" name="q" size="18" />
            <input type="submit" value="Go" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
          </form>
          <p class="searchtip" style="font-size: 90%">
            Enter search terms or a module, class or function name.
          </p>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <a href="introduction.html" title="Introduction"
             >previous</a> |
          <a href="cleaning.html" title="Cleaning sequence reads"
             >next</a> |
          <a href="genindex.html" title="General Index"
             >index</a>
            <br/>
            <a href="_sources/install.txt"
               rel="nofollow">Show Source</a>
        </div>

        <div class="right">
          
    <div class="footer">
        &copy; Copyright 2010, Jose Blanca.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.7.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>