

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Introduction &mdash; ngs_backbone v0.1 documentation</title>
    <link rel="stylesheet" href="_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="ngs_backbone v0.1 documentation" href="index.html" />
    <link rel="next" title="Installation" href="install.html" />
    <link rel="prev" title="ngs_backbone" href="index.html" /> 
  </head>
  <body>
    <div class="header-wrapper">
      <div class="header">
        <h1><a href="index.html">ngs_backbone v0.1 documentation</a></h1>
        <div class="rel">
          <a href="index.html" title="ngs_backbone"
             accesskey="P">previous</a> |
          <a href="install.html" title="Installation"
             accesskey="N">next</a> |
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="introduction">
<h1>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h1>
<p>ngs_backbone is an easy to use pipeline capable of doing NGS analyses in a reliable way.
This pipeline is capable of doing:</p>
<blockquote>
<div><ul class="simple">
<li>Read cleaning for sanger, 454, illumina and solid.</li>
<li>Mapping.</li>
<li>SNP calling and filtering.</li>
<li>ORF funtional annotation</li>
</ul>
</div></blockquote>
<p>The tool can be used to run any of these analyses independently of the others.
So it can be used only for read cleaning or only for SNP calling with SAM files obtained by using other tool.</p>
<p>The main design directives have been:</p>
<blockquote>
<div><ul class="simple">
<li>analyses reproducibility</li>
<li>ease of use</li>
<li>modularity</li>
<li>standard format output</li>
<li>use of third party tools when necessary.</li>
</ul>
</div></blockquote>
<p>By using ngs_backbone we can run a complete analysis in a reproducible way.
Every analysis parameter is configured in a text file and a log file is generated as the analysis progresses.
Also the analysis is easy to do, instead of running multiple scripts and programs only one command is required.</p>
<p>An application like that in a fast moving field, as the NGS is, has the risk of stagnate rapidly.
To avoid this pitfall special care has been taken in the design of the architecture of ngs_backbone.
Everything within the application is modular and several independent layers have been differentiated within the code to facilitate its maintenance.
In fact, before its public release ngs_backbone has gone through refactorings of several modules that have not affected the overall use of the application.</p>
</div>
<div class="section" id="third-party-software">
<h1>Third party software<a class="headerlink" href="#third-party-software" title="Permalink to this headline">¶</a></h1>
<p>ngs_backbone internally uses several external tools to run the analyses.
In most cases these tools are also free software and they are included in ngs_backbone, so you don&#8217;t need to worry about them.</p>
<table border="1" class="docutils">
<colgroup>
<col width="31%" />
<col width="69%" />
</colgroup>
<thead valign="bottom">
<tr><th class="head">analysis</th>
<th class="head">external tools required</th>
</tr>
</thead>
<tbody valign="top">
<tr><td><a class="reference internal" href="cleaning.html#clean-reads"><em>Cleaning sequence reads</em></a></td>
<td><a class="reference external" href="http://lucy.sourceforge.net/">lucy</a>, <a class="reference external" href="http://web.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">blast</a>, <a class="reference external" href="http://www.ncbi.nlm.nih.gov/VecScreen/UniVec.html">Univec</a> (database), <a class="reference external" href="http://compbio.dfci.harvard.edu/tgi/software/">mdust</a>, <a class="reference external" href="http://compbio.dfci.harvard.edu/tgi/software/">trimpoly</a></td>
</tr>
<tr><td><a class="reference internal" href="mapping.html"><em>Mapping</em></a></td>
<td><a class="reference external" href="http://bio-bwa.sourceforge.net/">bwa</a>, <a class="reference external" href="http://samtools.sourceforge.net/">samtools</a>, <a class="reference external" href="http://picard.sourceforge.net/index.shtml">picard</a></td>
</tr>
<tr><td><a class="reference internal" href="mapping.html#bam-realignment"><em>Bam realignment</em></a></td>
<td><a class="reference external" href="http://www.broadinstitute.org/gsa/wiki/index.php/The_Genome_Analysis_Toolkit">GATK</a></td>
</tr>
<tr><td><a class="reference internal" href="annotation.html#snp-calling"><em>SNP calling</em></a></td>
<td><a class="reference external" href="http://code.google.com/p/pysam/">pysam</a></td>
</tr>
<tr><td><a class="reference internal" href="annotation.html#orf-annotation"><em>ORF annotation</em></a></td>
<td><a class="reference external" href="http://estscan.sourceforge.net/">ESTScan</a></td>
</tr>
<tr><td><a class="reference internal" href="annotation.html#ortholog-annotation"><em>Ortholog annotation</em></a></td>
<td><a class="reference external" href="http://web.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">blast</a></td>
</tr>
<tr><td><a class="reference internal" href="annotation.html#description-annotation"><em>Description annotation</em></a></td>
<td><a class="reference external" href="http://web.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">blast</a></td>
</tr>
<tr><td><a class="reference internal" href="annotation.html#ssr-annotation"><em>Microsatellite annotation</em></a></td>
<td>sputnik</td>
</tr>
<tr><td><a class="reference internal" href="annotation.html#intron-annotation"><em>cDNA intron annotation</em></a></td>
<td><a class="reference external" href="http://web.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">blast</a>, <a class="reference external" href="http://emboss.sourceforge.net/">emboss</a></td>
</tr>
<tr><td><a class="reference internal" href="annotation.html#go-annotation"><em>GO annotation</em></a></td>
<td><a class="reference external" href="http://web.ncbi.nlm.nih.gov/blast/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">blast</a>, <a class="reference external" href="http://www.blast2go.org/">blast2go</a></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h1>
<p>The main ideas to consider when using ngs_backbone are: project and analysis.
A <em>project</em> is a directory (with its subdirectories) that includes a configuration file and all input and output files.
An <em>analysis</em> takes some inputs from the project and creates some outputs.
ngs_backbone knows where the input and output files are for every analysis because the project directory structure is the same for every project.
For instance the reads to clean are always in the directory /reads/original and the cleaned reads always are in /reads/cleaned/.
The main directories in this standard directory layout are:</p>
<blockquote>
<div><ul class="simple">
<li>backbone.conf</li>
<li>reads/raw</li>
<li>reads/cleaned</li>
<li>mapping/reference</li>
<li>mapping/bams</li>
<li>annotations/input</li>
<li>annotations/features</li>
</ul>
</div></blockquote>
<p>The configuration parameters required for every analysis are stored in the configuration file (backbone.conf) located in the project directory.
A sample of this configuration file is created by ngs_backbone when a new project is created.
This file should be tweaked to adapt the analyses to your requirements before running them.</p>
<p>ngs_backbone has only two main executables: backbone_create_project.py and backbone_analysis.py.
The first one is used to create a project from scratch and the second one to run all available analyses.</p>
<div class="section" id="creating-a-new-project">
<h2>Creating a new project<a class="headerlink" href="#creating-a-new-project" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><pre>$ backbone_create_project.py -p project_name
$ ls -l project_name
total 4
-rw-r--r-- 1 jose jose 2479 abr 16 09:30 ngs_backbone.conf</pre>
</div>
<p>This command will create a new directory named project_name with a file named ngs_backbone.conf in it.
These are the two hallmarks that define a ngs_backbone <em>project</em>, the directory and the configuration file.</p>
</div>
<div class="section" id="running-an-analysis">
<h2>Running an analysis<a class="headerlink" href="#running-an-analysis" title="Permalink to this headline">¶</a></h2>
<div class="highlight-python"><pre>$ backbone_analysis.py -a analysis_name</pre>
</div>
<p>This command will run the analysis on the data present in the project using the parameters found in the configuration file.
The output files will also be located inside the project.</p>
</div>
</div>
<div class="section" id="naming-conventions">
<span id="naming"></span><h1>Naming conventions<a class="headerlink" href="#naming-conventions" title="Permalink to this headline">¶</a></h1>
<p>The ngs_backbone usage is heavily based on directory and file name conventions.
If the input files are not located where ngs_backbone expects to find them or they have non-standard file names the analysis will fail.</p>
<table border="1" class="docutils">
<colgroup>
<col width="48%" />
<col width="52%" />
</colgroup>
<thead valign="bottom">
<tr><th class="head">file or directory type</th>
<th class="head">location</th>
</tr>
</thead>
<tbody valign="top">
<tr><td>configuration file</td>
<td>backbone.conf</td>
</tr>
<tr><td>log file</td>
<td>backbone.log</td>
</tr>
<tr><td>raw reads</td>
<td>reads/raw/</td>
</tr>
<tr><td>clean reads</td>
<td>reads/cleaned/</td>
</tr>
<tr><td>assemblies</td>
<td>assembly/</td>
</tr>
<tr><td>mappings</td>
<td>mapping/</td>
</tr>
<tr><td>mapping output</td>
<td>mapping/bams/</td>
</tr>
<tr><td>annotations</td>
<td>annotations/</td>
</tr>
<tr><td>annotation input</td>
<td>annotations/input/</td>
</tr>
<tr><td>annotation output</td>
<td>annotation/features/</td>
</tr>
<tr><td>error logs</td>
<td>backbone_errors/</td>
</tr>
</tbody>
</table>
<p>Sequence files should be fasta or sanger fastq, fasta for the ones without quality and fastq for the ones with quality.
The file extension should reflect the sequence file format, fasta and sfastq.</p>
<p>The sequence file names for the reads should define several tags: lb (library), sm (sample) and pl (platform/technology).
This convention is required if a mapping analysis is to be done with these sequence files.
These tags follow the sam file header <a class="reference external" href="http://samtools.sourceforge.net/SAM-1.3.pdf">specification</a>.
An example of some valid sequence file names is:</p>
<div class="highlight-python"><pre>$ ls reads/raw/
lb_mos.pl_illumina.sm_mos.sfastq  lb_pep.pl_illumina.sm_pep.sfastq
lb_mu16.pl_454.sm_mu16.sfastq     lb_upv196.pl_454.sm_upv196.sfastq</pre>
</div>
</div>
<div class="section" id="available-analyses">
<h1>Available analyses<a class="headerlink" href="#available-analyses" title="Permalink to this headline">¶</a></h1>
<p>The available analyses are:</p>
<table border="1" class="docutils">
<colgroup>
<col width="45%" />
<col width="55%" />
</colgroup>
<thead valign="bottom">
<tr><th class="head">analysis</th>
<th class="head">description</th>
</tr>
</thead>
<tbody valign="top">
<tr><td><a class="reference internal" href="cleaning.html#clean-reads"><em>Cleaning sequence reads</em></a></td>
<td>sequence reads cleaning</td>
</tr>
<tr><td><a class="reference internal" href="mapping.html"><em>Mapping</em></a></td>
<td>bwa read mapping against a reference sequence</td>
</tr>
<tr><td><a class="reference internal" href="mapping.html#bam-realignment"><em>Bam realignment</em></a></td>
<td>GATK bam realignment</td>
</tr>
<tr><td><a class="reference internal" href="annotation.html#snp-calling"><em>SNP calling</em></a></td>
<td>SNP annotation from a bam file</td>
</tr>
<tr><td><a class="reference internal" href="annotation.html#orf-annotation"><em>ORF annotation</em></a></td>
<td>ESTScan ORF annotation</td>
</tr>
<tr><td><a class="reference internal" href="annotation.html#ortholog-annotation"><em>Ortholog annotation</em></a></td>
<td>reciprocal blast based ortholog annotation</td>
</tr>
<tr><td><a class="reference internal" href="annotation.html#description-annotation"><em>Description annotation</em></a></td>
<td>description blast based annotation</td>
</tr>
<tr><td><a class="reference internal" href="annotation.html#ssr-annotation"><em>Microsatellite annotation</em></a></td>
<td>microsatellite sputnik based annotation</td>
</tr>
<tr><td><a class="reference internal" href="annotation.html#intron-annotation"><em>cDNA intron annotation</em></a></td>
<td>est2genome cDNA based annotation</td>
</tr>
<tr><td><a class="reference internal" href="annotation.html#go-annotation"><em>GO annotation</em></a></td>
<td>blast2go  annotation</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="parallel-operation">
<h1>Parallel operation<a class="headerlink" href="#parallel-operation" title="Permalink to this headline">¶</a></h1>
<p>Running ngs_backbone with multiple subprocesses is as easy as setting the configuration option threads to True.
ngs_backbone will run will as many subprocesses as cpu cores are found in the computer.
Also the threads option can be set to an integer and ngs_backbone will run with as many subprocess as indicated.</p>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          <h3>Table Of Contents</h3>
          <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="#third-party-software">Third party software</a></li>
<li class="toctree-l1"><a class="reference internal" href="#usage">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#creating-a-new-project">Creating a new project</a></li>
<li class="toctree-l2"><a class="reference internal" href="#running-an-analysis">Running an analysis</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#naming-conventions">Naming conventions</a></li>
<li class="toctree-l1"><a class="reference internal" href="#available-analyses">Available analyses</a></li>
<li class="toctree-l1"><a class="reference internal" href="#parallel-operation">Parallel operation</a></li>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="cleaning.html">Cleaning sequence reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="assembly.html">Mira assembly</a></li>
<li class="toctree-l1"><a class="reference internal" href="mapping.html">Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="mapping.html#bam-realignment">Bam realignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="annotation.html">Annotation</a></li>
<li class="toctree-l1"><a class="reference internal" href="snv_filters.html">Snv filters</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="seqio.html">seq_io</a></li>
<li class="toctree-l1"><a class="reference internal" href="seqio.html#seqio-supported-formats">Seqio supported formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="architecture.html">Architecture</a></li>
<li class="toctree-l1"><a class="reference internal" href="download.html">Download</a></li>
</ul>

          <h3 style="margin-top: 1.5em;">Search</h3>
          <form class="search" action="search.html" method="get">
            <input type="text" name="q" size="18" />
            <input type="submit" value="Go" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
          </form>
          <p class="searchtip" style="font-size: 90%">
            Enter search terms or a module, class or function name.
          </p>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <a href="index.html" title="ngs_backbone"
             >previous</a> |
          <a href="install.html" title="Installation"
             >next</a> |
          <a href="genindex.html" title="General Index"
             >index</a>
            <br/>
            <a href="_sources/introduction.txt"
               rel="nofollow">Show Source</a>
        </div>

        <div class="right">
          
    <div class="footer">
        &copy; Copyright 2010, Jose Blanca.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.7.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>