

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Snv filters &mdash; ngs_backbone v0.1 documentation</title>
    <link rel="stylesheet" href="_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="ngs_backbone v0.1 documentation" href="index.html" />
    <link rel="next" title="Tutorials" href="tutorials.html" />
    <link rel="prev" title="Annotation" href="annotation.html" /> 
  </head>
  <body>
    <div class="header-wrapper">
      <div class="header">
        <h1><a href="index.html">ngs_backbone v0.1 documentation</a></h1>
        <div class="rel">
          <a href="annotation.html" title="Annotation"
             accesskey="P">previous</a> |
          <a href="tutorials.html" title="Tutorials"
             accesskey="N">next</a> |
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="snv-filters">
<h1>Snv filters<a class="headerlink" href="#snv-filters" title="Permalink to this headline">¶</a></h1>
<p>Once we have called the  snvs with the annotate_snv backbone analysis, we can filter those snvs.
This analysis is performed by the filter_snv analysis.</p>
<p>The configuration of the snv filtering is handled by the snv_filters section of the backbone.conf file. For each filter that we want to apply we have to create a new subsection like:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="p">[</span><span class="s">&#39;snv_filters&#39;</span><span class="p">]</span>
<span class="p">[[</span><span class="s">&#39;filter1&#39;</span><span class="p">]]</span>
<span class="n">name</span> <span class="o">=</span> <span class="s">&#39;close_to_intron&#39;</span>
<span class="n">distance</span> <span class="o">=</span> <span class="mi">30</span>
<span class="n">use</span>  <span class="o">=</span> <span class="bp">True</span>
</pre></div>
</div>
<p>All the snv filters have a boolean configuration parameter named <em>use</em>, when this parameter is set to False the filter will not be used even if it is declared in the configuration file.</p>
<p>By default only kind filtering is activated although, as an example, other filters are configured with its use parameter set to False. To use any of these filters change the <em>use</em> parameter to True in each of the filter sections:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">use</span> <span class="o">=</span> <span class="bp">True</span>
</pre></div>
</div>
<p>None of the  filters remove the snv, instead it adds a flag to the snv depending on the result of the filter. The information regarding the snv filtering will be written in the vcf file.</p>
<div class="section" id="available-filters">
<h2>Available filters<a class="headerlink" href="#available-filters" title="Permalink to this headline">¶</a></h2>
<p>The configuration parameters for each available filter is described here.</p>
<div class="section" id="close-to-intron">
<h3>Close to intron<a class="headerlink" href="#close-to-intron" title="Permalink to this headline">¶</a></h3>
<p>The snvs close to an intron will be filtered by this filter. To use it you have to annotate the introns before using the <a class="reference internal" href="annotation.html#intron-annotation"><em>cDNA intron annotation</em></a> backbone analysis.</p>
<p>Use this filter if you want to filter snvs that are closer than X nucleotides to an intron.
The configuration parameter for this filter is the distance between the snv and the intron:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">name</span>     <span class="o">=</span> <span class="s">&#39;close_to_intron&#39;</span>
<span class="n">distance</span> <span class="o">=</span> <span class="mi">30</span>
</pre></div>
</div>
</div>
<div class="section" id="close-to-snv">
<h3>Close to snv<a class="headerlink" href="#close-to-snv" title="Permalink to this headline">¶</a></h3>
<p>Use this filter if you want to filter out snvs that are close to another snv.</p>
<p>This filter only have a parameter, distance between snvs:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">name</span>     <span class="o">=</span> <span class="s">&#39;close_to_snv&#39;</span>
<span class="n">distance</span> <span class="o">=</span> <span class="mi">60</span>
</pre></div>
</div>
</div>
<div class="section" id="close-to-limit">
<h3>Close to limit<a class="headerlink" href="#close-to-limit" title="Permalink to this headline">¶</a></h3>
<p>Use this filter if you want to filter snvs that are close to the edges of the sequence.</p>
<p>The parameter to set for this filter is:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">name</span>     <span class="o">=</span> <span class="s">&#39;close_to_limit&#39;</span>
<span class="n">distance</span> <span class="o">=</span> <span class="mi">60</span>
</pre></div>
</div>
</div>
<div class="section" id="high-variable-region">
<h3>High variable region<a class="headerlink" href="#high-variable-region" title="Permalink to this headline">¶</a></h3>
<p>It filters out all the snvs found in a sequence with a high variability. The variability is measured as the number of snvs per base.</p>
<p>Two parameters can be configured for this filter, the variability threshold and the window. If the window is not given the complete sequence is used to calculate the variability.</p>
<dl class="docutils">
<dt>::</dt>
<dd>name            = &#8216;high_variable_region&#8217;
max_variability = 0.04
window          = None  #(All the sequence)</dd>
</dl>
</div>
<div class="section" id="more-frequent-allele">
<h3>More frequent allele<a class="headerlink" href="#more-frequent-allele" title="Permalink to this headline">¶</a></h3>
<p>With this filter we can filter out the filters in which the most frequent allele has an allelic frequency above the given threshold. This filter can be used taking into account all the alleles or only the ones found in a subset of libraries, samples or readgroups.</p>
<p>To configure this filter with a threshold of 0.8 in a library named test_library we would do:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">maf</span>        <span class="o">=</span> <span class="s">&#39;maf&#39;</span>
<span class="n">frequency</span>  <span class="o">=</span> <span class="mf">0.8</span>
<span class="n">group_kind</span> <span class="o">=</span> <span class="s">&#39;libraries&#39;</span>
<span class="n">groups</span>     <span class="o">=</span> <span class="p">[</span><span class="s">&#39;test_library&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="section" id="kind-filter">
<h3>Kind filter<a class="headerlink" href="#kind-filter" title="Permalink to this headline">¶</a></h3>
<p>The backbone snv caller annotates SNPs and indels, with this filter, for instance, we can filter out the snvs that are not SNPs.
The snv caller sets a kind for each snv: SNP, INDEL or COMPLEX. You can filter by:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="s">&#39;SNV&#39;</span><span class="p">,</span> <span class="s">&#39;INVARIANT&#39;</span><span class="p">,</span> <span class="s">&#39;COMPLEX&#39;</span>
</pre></div>
</div>
<p>This filter only has one option to configure:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">name</span> <span class="o">=</span> <span class="s">&#39;by_kind&#39;</span>
<span class="n">kind</span> <span class="o">=</span> <span class="s">&#39;SNV&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="cap-enzymes">
<h3>Cap enzymes<a class="headerlink" href="#cap-enzymes" title="Permalink to this headline">¶</a></h3>
<p>We can filter out the snvs that will not be detectable by using restriction enzymes. We consider a snv detectable if the two most abundant alleles have different restriction patterns.
This filter uses remap from EMBOSS to do the restriction mapping. We can choose to use all the restriction enzymes from the remap database or only a small subset of cheap and easily available enzymes. The parameter to configure is:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">name</span>        <span class="o">=</span> <span class="s">&#39;cap_enzyme&#39;</span>
<span class="n">all_enzymes</span> <span class="o">=</span> <span class="bp">True</span>
</pre></div>
</div>
</div>
<div class="section" id="sequence-filter">
<h3>Sequence filter<a class="headerlink" href="#sequence-filter" title="Permalink to this headline">¶</a></h3>
<p>It filters out the snv that do not belong to one of the given sequences. The sequences should be provided as a list in a file with one sequence name per line.</p>
<p>The filter requires a parameter with the path to the sequence names file:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">name</span>      <span class="o">=</span> <span class="s">&#39;ref_not_in_list&#39;</span>
<span class="n">list_path</span> <span class="o">=</span> <span class="s">&#39;/path/to/file/with/seq/names&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="variable-in-group">
<h3>Variable in group<a class="headerlink" href="#variable-in-group" title="Permalink to this headline">¶</a></h3>
<p>With this filter we can select SNVs variables in some specific samples, libraries or read groups. For instance we can look for SNVs variables in the sample1.</p>
<p>It can work with libraries, samples or read_groups. The kind of grouping used is set in the variable group_kind.</p>
<p>The groups are a list (for instance a list of samples). By default it will look for SNVs variable in the union of these samples. If we ask for SNVs variable in sample_1 and sample_2 it will consider SNVs with different alleles between 1 and 2 and also SNVs with two alleles in sample_1 but not variable in sample_2. If we want all the samples to be variable independently of each other we could set the parameters in_union to False and in_all_groups to True.</p>
<p>To configure the filter the kind of group to use (libraries, samples or read_groups) should be set. Also a list with the group names should be given. A configuration to look for SNVs variable in the libraries lib1 and lib2 would be:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">name</span>        <span class="o">=</span> <span class="s">&#39;is_variable&#39;</span>
<span class="n">unique_name</span> <span class="o">=</span> <span class="s">&#39;is_variable_in_lb&#39;</span>
<span class="n">group_kind</span>  <span class="o">=</span> <span class="s">&#39;libraries&#39;</span>
<span class="n">groups</span>      <span class="o">=</span> <span class="p">[</span><span class="s">&#39;lib1&#39;</span><span class="p">,</span> <span class="s">&#39;lib2&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Several filters of this kind can be set up in the configuration file, to distinguish them a unique_name option with a  should be also set up in the configuration for each filter.</p>
</div>
<div class="section" id="not-variable-in-group">
<h3>Not variable in group<a class="headerlink" href="#not-variable-in-group" title="Permalink to this headline">¶</a></h3>
<p>With this filter we can select SNVs not variable in some specific samples, libraries or read groups. For instance we can look for SNVs not variables in the sample_1.</p>
<p>It can work with libraries, samples or read_groups. The kind of grouping used is set in the variable group_kind.</p>
<p>The groups are a list (for instance a list of samples). If we ask for SNVs not variable in sample_1 and sample_2 it will consider SNVs that are not variable in all samples (they all should have the same allele). If we want the samples to be not variable but we want to allow to have different alleles between them we could use the option in_union=False.</p>
<p>To configure the filter the kind of group to use (libraries, samples or read_groups) should be set. Also a list with the group names should be given. A configuration to look for SNVs not variable in the libraries lib1 and lib2 would be:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">name</span>        <span class="o">=</span> <span class="s">&#39;is_not_variable&#39;</span>
<span class="n">unique_name</span> <span class="o">=</span> <span class="s">&#39;not_is_variable_in_lb&#39;</span>
<span class="n">group_kind</span>  <span class="o">=</span> <span class="s">&#39;libraries&#39;</span>
<span class="n">groups</span>      <span class="o">=</span> <span class="p">[</span><span class="s">&#39;lib1&#39;</span><span class="p">,</span> <span class="s">&#39;lib2&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>Several filters of this kind can be set up in the configuration file, to distinguish them a unique_name option with a  should be also set up in the configuration for each filter.</p>
</div>
<div class="section" id="unique-contiguous">
<h3>Unique contiguous<a class="headerlink" href="#unique-contiguous" title="Permalink to this headline">¶</a></h3>
<p>With this filter you can filter out snvs that are in regions that seem to be duplicated or that are not contiguous.</p>
<p>This filter have 4 configurable options:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">name</span>               <span class="o">=</span> <span class="s">&#39;uniq_contiguous&#39;</span>
<span class="n">distance</span>           <span class="o">=</span> <span class="s">&#39;distance from each side of the snv to select a region&#39;</span>
<span class="n">genomic_db</span>         <span class="o">=</span> <span class="s">&#39;/path/to/the/seq/fasta/file&#39;</span>
<span class="n">genomic_seqs_fpath</span> <span class="o">=</span> <span class="s">&#39;/path/to/the/seq/blast/db&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="minimun-number-of-groups">
<h3>Minimun number of groups<a class="headerlink" href="#minimun-number-of-groups" title="Permalink to this headline">¶</a></h3>
<p>It filters out the SNVs that not supported by enough number of groups. For instance we could filter the SNVs read in less than 4 samples.</p>
<p>The options are:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">name</span>       <span class="o">=</span> <span class="s">&#39;min_groups&#39;</span>
<span class="n">min_groups</span> <span class="o">=</span> <span class="mi">4</span>
<span class="n">group_kind</span> <span class="o">=</span> <span class="s">&#39;samples&#39;</span>
</pre></div>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          <h3>Table Of Contents</h3>
          <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#third-party-software">Third party software</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#usage">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#naming-conventions">Naming conventions</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#available-analyses">Available analyses</a></li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html#parallel-operation">Parallel operation</a></li>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="cleaning.html">Cleaning sequence reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="assembly.html">Mira assembly</a></li>
<li class="toctree-l1"><a class="reference internal" href="mapping.html">Mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="mapping.html#bam-realignment">Bam realignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="annotation.html">Annotation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Snv filters</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#available-filters">Available filters</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="seqio.html">seq_io</a></li>
<li class="toctree-l1"><a class="reference internal" href="seqio.html#seqio-supported-formats">Seqio supported formats</a></li>
<li class="toctree-l1"><a class="reference internal" href="architecture.html">Architecture</a></li>
<li class="toctree-l1"><a class="reference internal" href="download.html">Download</a></li>
</ul>

          <h3 style="margin-top: 1.5em;">Search</h3>
          <form class="search" action="search.html" method="get">
            <input type="text" name="q" size="18" />
            <input type="submit" value="Go" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
          </form>
          <p class="searchtip" style="font-size: 90%">
            Enter search terms or a module, class or function name.
          </p>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <a href="annotation.html" title="Annotation"
             >previous</a> |
          <a href="tutorials.html" title="Tutorials"
             >next</a> |
          <a href="genindex.html" title="General Index"
             >index</a>
            <br/>
            <a href="_sources/snv_filters.txt"
               rel="nofollow">Show Source</a>
        </div>

        <div class="right">
          
    <div class="footer">
        &copy; Copyright 2010, Jose Blanca.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.7.
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>